mender-update/daemon/states.cpp

```diff
@@ -289,24 +289,87 @@ void UpdateDownloadState::ParseArtifact(Context &ctx, sm::EventPoster<StateEvent
 }
 
 SendStatusUpdateState::SendStatusUpdateState(
-	optional::optional<deployments::DeploymentStatus> status, FailureMode mode) :
+	optional::optional<deployments::DeploymentStatus> status) :
 	status_(status),
-	mode_(mode) {
+	mode_(FailureMode::Ignore) {
+}
+
+SendStatusUpdateState::SendStatusUpdateState(
+	optional::optional<deployments::DeploymentStatus> status,
+	events::EventLoop &event_loop,
+	int retry_interval_seconds) :
+	status_(status),
+	mode_(FailureMode::RetryThenFail),
+	retry_(Retry {http::ExponentialBackoff(chrono::seconds(retry_interval_seconds)), event_loop}) {
+	// MEN-2676: Cap at 10 retries.
+	retry_->backoff.SetTryCount(10);
+}
+
+void SendStatusUpdateState::SetSmallestWaitInterval(chrono::milliseconds interval) {
+	if (retry_) {
+		retry_->backoff.SetSmallestInterval(interval);
+	}
 }
 
 void SendStatusUpdateState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	// Reset this every time we enter the state, which means a new round of retries.
+	if (retry_) {
+		retry_->backoff.Reset();
+	}
+
+	DoStatusUpdate(ctx, poster);
+}
+
+void SendStatusUpdateState::DoStatusUpdate(Context &ctx, sm::EventPoster<StateEvent> &poster) {
 	assert(ctx.deployment_client);
 	assert(ctx.deployment.state_data);
 
-	auto result_handler = [this, &poster](const error::Error &err) {
+	log::Info("Sending status update to server");
+
+	auto result_handler = [this, &ctx, &poster](const error::Error &err) {
 		if (err != error::NoError) {
 			log::Error("Could not send deployment status: " + err.String());
+
 			switch (mode_) {
 			case FailureMode::Ignore:
 				break;
-			case FailureMode::Fail:
-			case FailureMode::RetryThenFail: // MEN-6573: Handle this
-				poster.PostEvent(StateEvent::Failure);
+			case FailureMode::RetryThenFail:
+				if (err.code
+					== deployments::MakeError(deployments::DeploymentAbortedError, "").code) {
```

Comment:
Can't we do:
```suggestion
				if (err.code == deployments::DeploymentAbortedError) {
```
? 

Maybe I am missing something